pkg/services/gitopswriter/gitopswriter.go

```diff
@@ -0,0 +1,213 @@
+package gitopswriter
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+
+	"github.com/fluxcd/go-git-providers/gitprovider"
+	"github.com/weaveworks/weave-gitops/pkg/git"
+	"github.com/weaveworks/weave-gitops/pkg/gitproviders"
+	"github.com/weaveworks/weave-gitops/pkg/logger"
+	"github.com/weaveworks/weave-gitops/pkg/models"
+	"github.com/weaveworks/weave-gitops/pkg/osys"
+	"github.com/weaveworks/weave-gitops/pkg/services/automation"
+	"github.com/weaveworks/weave-gitops/pkg/services/gitrepo"
+
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	AddCommitMessage    = "Add application manifests"
+	RemoveCommitMessage = "Remove application manifests"
+)
+
+var _ GitOpsDirectoryWriter = &GitOpsDirectoryWriterSvc{}
+
+type GitOpsDirectoryWriter interface {
+	AddApplication(ctx context.Context, app models.Application, clusterName string, autoMerge bool) error
+	RemoveApplication(ctx context.Context, app models.Application, clusterName string, autoMerge bool) error
+}
+
+type GitOpsDirectoryWriterSvc struct {
```

Comment:
Nit: does this need to be exported? The constructor seems to encapsulate everything we need to add to the struct.

Reply:
Good point.

Reply:
BTW, GitHub doesn't want to let me reply to your question above about the moved tests. This is what I tried to say:

Yes, they were moved but the invocation was changed from calling app.Add to calling the gitopswriter.AddApplication directly. Only the tests that check parameter validation (and the original "new directory structure" tests) are left in add_test.go in app. The tests that were purely concerned with manifest generation moved to the automation service and the ones that tested "adding" but not add parameters got moved to the gitopswriter.